File: DemoApp/DemoApp/BannersDetailViewController.m
//
//  BannersDetailViewController.m
//  DemoApp
//
//  Created by Glispa GmbH on 04/12/15.
//  Copyright © 2015 Glispa GmbH All rights reserved.
//

#import "BannersDetailViewController.h"
#import "AppControllerManagement.h"

@interface BannersDetailViewController ()

@property(weak, nonatomic) IBOutlet UISwitch *autoupdateSwitch;
@property(strong, nonatomic) AVOBannerView *bannerView;
@end

@implementation BannersDetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self loadBanner];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.bannerView resumeAutoUpdate];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.bannerView pauseAutoUpdate];
}

- (void)loadBanner {
    [self.bannerView stop];
    [self.bannerView removeFromSuperview];
    self.bannerView = nil;
    
    __weak typeof(self) weakSelf = self;
    [AvocarrotSDK.sharedSDK loadBannerWithSize:self.bannerSize
                                      adUnitId:self.adUnitId
                                       success:^(AVOBannerView * _Nonnull bannerAd) {
                                           __strong __typeof__(self) sSelf = weakSelf;
                                           [sSelf addBannerOnScreen:bannerAd];
                                       }
                                       failure:^(AVOError * _Nonnull error) {
                                           NSLog(@"Banner loading error: %@", [error localizedDescription]);
                                       }];
}

#pragma mark - Actions

- (IBAction)autoupdateSwitched:(UISwitch *)switcher {
    if (self.bannerView) {
        self.bannerView.autoUpdate = switcher.isOn;
    }
}

#pragma mark - Banner ad

- (void)addBannerOnScreen:(AVOBannerView *)newBanner {
    [self.bannerView stop];
    [self.bannerView removeFromSuperview];
    self.bannerView = newBanner;
    
    [self.view addSubview:self.bannerView];
    self.bannerView.autoUpdate = self.autoupdateSwitch.isOn;
    self.bannerView.translatesAutoresizingMaskIntoConstraints = NO;
    
    NSDictionary *views = @{@"bannerView" : self.bannerView,
                            @"superview" : self.view};
    NSDictionary *metrics = @{@"width" : @(self.bannerView.frame.size.width),
                              @"height" : @(self.bannerView.frame.size.height)};
    
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:[bannerView(==height)]-20-|" options:0 metrics:metrics views:views]];
    
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:[bannerView(==width)]" options:0 metrics:metrics views:views]];
    
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:[superview]-(<=1)-[bannerView]" options:NSLayoutFormatAlignAllCenterX metrics:metrics views:views]];
}

@end


@interface BannersDetailViewController (AppManagement) <AppControllerManagement>

@end

@implementation BannersDetailViewController (AppManagement)

- (void)applyUserData:(NSDictionary <NSString *, NSObject *> *)userData {
    if ([userData[@"adUnitId"] isKindOfClass:[NSString class]]) {
        self.adUnitId = (NSString *)userData[@"adUnitId"];
    }
    
    if ([userData[@"bannerType"] isKindOfClass:[NSNumber class]]) {
        self.bannerSize = (AVOBannerViewSize)[(NSNumber *)userData[@"bannerType"] integerValue];
    }
}

@end

File: DemoApp/DemoAppSwift/BannersDetailViewController.swift
//
//  BannersDetailViewController.swift
//  DemoApp
//
//  Created by Glispa GmbH on 19/05/16.
//  Copyright © 2016 Glispa GmbH. All rights reserved.
//

import UIKit
import AvocarrotBanner

class BannersDetailViewController: UIViewController {

    @IBOutlet weak var autoupdateSwitch: UISwitch!

    fileprivate var bannerView: AVOBannerView?
    public var bannerSize: AVOBannerViewSize!
    public var adUnitId: String!

    override func viewDidLoad() {
        super.viewDidLoad()
        loadBanner()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        bannerView?.resumeAutoUpdate()
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        bannerView?.pauseAutoUpdate()
    }

    // MARK: - Public

    public func loadBanner() {
        AvocarrotSDK.shared.loadBanner(with: bannerSize, adUnitId: adUnitId,
                                       success: {[weak self] (bannerAd) in
                                            self?.addBannerOnScreen(newBanner: bannerAd)
                                        },
                                       failure: { (error) in
                                            print("Banner loading error: \(error.localizedDescription)")
                                       })
    }

    // MARK: - Actions

    @IBAction func autoupdateSwitched(_ sender: UISwitch) {
        bannerView?.autoUpdate = sender.isOn
    }
    
    // MARK: - Private
    private func addBannerOnScreen(newBanner: AVOBannerView) {
        bannerView?.stop()
        bannerView?.removeFromSuperview()
        bannerView = newBanner
        
        view.addSubview(newBanner)
        newBanner.translatesAutoresizingMaskIntoConstraints = false
        
        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat:
            "V:[bannerView(\(newBanner.frame.size.height))]-20-|",
            options: NSLayoutFormatOptions(),
            metrics: nil,
            views: ["bannerView": newBanner]))
        
        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat:
            "H:[bannerView(\(newBanner.frame.size.width))]",
            options: .alignAllCenterX,
            metrics: nil,
            views: ["bannerView": newBanner]))
        
        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat:
            "V:[superview]-(<=1)-[bannerView]",
                                                           options: .alignAllCenterX,
                                                           metrics: nil,
                                                           views: ["bannerView": newBanner,
                                                                   "superview": view]))

    }
}

extension BannersDetailViewController: AppControllerManagement {
    func applyUserData(_ userData: [String:Any]) {
        if let type = userData["bannerType"] as? AVOBannerViewSize {
            bannerSize = type
        }

        if let adId = userData["adUnitId"] as? String {
            adUnitId = adId
        }        
    }
}
